File: apps/next/maze-runner/components/maze.js
export default (maze = {}) => {
    let m = {
        wallChar: (maze.wallChar ? maze.wallChar : "#"),
        startChar: (maze.startChar ? maze.startChar : "s"),
        endChar: (maze.endChar ? maze.endChar : "e"),
        openChar: (maze.openChar ? maze.openChar : " "),
        layout: (maze.layout ? maze.layout.trim().split('\n').map((i)=> i.trim().split("")) : []),
        height: (maze.layout ? maze.layout.trim().split('\n').length : 0),
        width: (maze.layout ? maze.layout.trim().split('\n')[0].length : 0),
        
        buildNew(h=10, w=10, mazeType="h") {
            let openPoints = [],
                newLayout = [],
                choices = [this.openChar, this.openChar, this.wallChar],
                s = [0, 0],
                e = [0, 0];
                this.height = h;
                this.width = w;
            for (let i = 0; i < h; i++){
                newLayout.push([]);
                for (let j = 0; j < w; j++){
                    newLayout[i].push([i, j]);
                }
            }
            for (let x = 0; x < newLayout.length; x++){ 
                for (let y = 0; y < newLayout[x].length; y++) {
                    let p = newLayout[x][y];
                    if (p[0] == 0 || p[1] == 0 || p[0] == this.height - 1 || p[1] == this.width - 1) {
                        newLayout[x][y] = this.wallChar;
                    } else {
                        openPoints.push([x, y]);
                    }
                }
            }
            this.layout = newLayout;
            if (mazeType == "h") {
                s = [1, Math.floor(Math.random() * (this.width - 2)) + 1];
                openPoints = openPoints.filter((i) => i[0] !== s[0] || i[1] !== s[1]);
                e = [this.height - 2, Math.floor(Math.random() * (this.width - 2)) + 1];
                openPoints = openPoints.filter((i) => i[0] !== e[0] || i[1] !== e[1]);
            } else if (mazeType == "v") {
                s = [Math.floor(Math.random() * (this.height - 2)) + 1, 1];
                openPoints = openPoints.filter((i) => i[0] !== s[0] || i[1] !== s[1]);
                e = [Math.floor(Math.random() * (this.height - 2)) + 1, this.width - 2];
                openPoints = openPoints.filter((i) => i[0] !== e[0] || i[1] !== e[1]);
            } else if (mazeType == "r") {
                while (s[0] == e[0] && s[1] == e[1]) {
                    s = [Math.floor(Math.random() * (this.height - 2)) + 1, Math.floor(Math.random() * (this.width - 2)) + 1];
                    e = [Math.floor(Math.random() * (this.height - 2)) + 1, Math.floor(Math.random() * (this.width - 2)) + 1];
                }
                openPoints = openPoints.filter((i) => i[0] !== s[0] || i[1] !== s[1]);
                openPoints = openPoints.filter((i) => i[0] !== e[0] || i[1] !== e[1]);

            } else {
                throw "Incorrect Maze type: try h, v, or r.";
            }
            this.layout[s[0]][s[1]] = this.startChar;
            this.layout[e[0]][e[1]] = this.endChar;

            openPoints.map((i)=> {
                this.layout[i[0]][i[1]] = choices[Math.floor(Math.random() * choices.length)];
            });
            
            return this.layout;
        },

        typeInfo() {
            console.log(`    start point: ${this.startChar}\n    end point: ${this.endChar}\n    open spaces: ${this.openChar}\n    wall type: ${this.wallChar}\n    size: ${this.height} x ${this.width}`);
        },

        viewLayout() {
            return console.log(this.layout.map((i)=> i.join("")).join('\n'));
        }
    }
    if (maze.layout == undefined) {
        let h = (maze.build ? maze.build[0] : 10);
        let w = (maze.build ? maze.build[1] : 10);
        let t = (maze.buildType ? maze.buildType : "h");
        m.buildNew(h, w, t);
    }
    return m;
}

File: apps/flask/modules/maze.py
import random

class Maze:
	def __init__ (self, layout=None, start_char="s", end_char="e", wall_char="#", open_char=" ", build=[(10, 10), "h"]):
		self.wall_char = wall_char
		self.start_char = start_char
		self.end_char = end_char
		self.open_char = open_char
		self.layout = layout
		if layout:
			self.width = len(layout[0])
			self.height = len(layout)
		else:
			self.build_new(build[0][0], build[0][1], build[1])

	def build_new(self, width=None, height=None, maze_type = "h"):
		if height == None:
			height = self.height 
		else:
			self.height = height
		if width == None:
			width = self.width 
		else:
			self.width = width
		self.layout = [[(h, w) for w in range(width)] for h in range(height)]
		open_points = []
		for x in range(len(self.layout)):
			for y in range(len(self.layout[x])):
				p = self.layout[x][y]
				if p[0] == 0 or p[1] == 0 or p[0] == height-1 or p[1] == width-1:
					self.layout[x][y] = self.wall_char
				else:
					open_points.append((x,y))
		if maze_type == "h":
			s= (1, random.choice(range(1, width-1)))
			del open_points[open_points.index(s)]
			e = (height-2, random.choice(range(1, width-1)))
			del open_points[open_points.index(e)]
		elif maze_type == "v":
			s = (random.choice(range(1, height-1)), 1)
			del open_points[open_points.index(s)]
			e = (random.choice(range(1, height-1)), width-2)
			del open_points[open_points.index(e)]
		elif maze_type == "r":
			s = random.choice(open_points)
			del open_points[open_points.index(s)]
			e = random.choice(open_points)
			del open_points[open_points.index(e)]
		else:
			raise ValueError("Incorrect Maze type. (try h, v, or r)")
		self.layout[s[0]][s[1]] = self.start_char
		self.layout[e[0]][e[1]] = self.end_char
		for i in open_points:
			self.layout[i[0]][i[1]] = random.choice([self.open_char, self.open_char, self.wall_char])
		return self.layout

	def type_info(self):
		print(f"	start point: {self.start_char}\n	end point: {self.end_char}\n	open spaces: {self.open_char}	wall type: {self.wall_char}\n	size: {maze.height} x {maze.width}")

	def view_layout(self):
		x =""
		for i in self.layout:
			x += "".join(i)
			x += "\n"
		print(x)
		return x